Add env to field callback

chukinas opened the pull request:
This PR does two different things:
1. (refactor) Replace `Module.eval_quoted` calls by using a `@before_compile` callback. You'd mentioned you were open to strategies for doing that [here](https://github.com/ejpcmac/typed_struct/issues/21#issuecomment-1003442610).
2. (feature) Add the current Macro.Env as a fourth param to the field callback. The addition of `field/4` is fully backwards-compatible with `field/3`.

What do you think?

chukinas committed:
refactor: remove Module.eval_quoted part 2

1. Move the handling of the "after definition" callbacks to the
   before_compile stage. This allows us to get rid of this particular
   Module.eval_quoted.
2. Because the setup and teardown of module attributes is now spread
   out over two macros, it makes sense to extract and then loop over an
   attrs list.

chukinas committed:
refactor: remove Module.eval_quoted part 3

Refactor the plugin field callbacks
1. Accumulate each field's name, type, opts and current ENV in a new
   module attribute (Is there an opportunity to combine this with
   @ts_fields later?).
2. In before_compile macro, loop over fields and plugins to generate an
   ast of all the plugin field callbacks.
3. Inject this ast and the after_definition block.

chukinas committed:
refactor: remove Module.eval_quoted part 1

1. remove Module/eval_quoted from typedstruct/2
2. move the "create lexical scope" anon function from __typedstruct__/2
   into the "no module" case of typedstruct/2. (There's no need to do it
   the "module" case, because modules already create their own lexical
   scope.

chukinas committed:
feat: make `env` available to field callback

1. Add TypedStruct.Plugin.field/4 callback. This is the same as the old
   field/3, but with the current env as the 4th arg.
2. Soft-deprecate TypedStruct.Plugin.field/3
3. `use`-ing Plugin implements a default, overrideable field/4
4. Add a default implementation of after_definition/1 so the docs are
   consistent. Now they all follow the GenServer pattern - "defaults for
   all; override them if you want"

chukinas committed:
test: add tests for new field/4 env

ejpcmac commented:
Hello, thank you for your PR! I just took a first glance on it, seems pretty good :slightly_smiling_face: I’ll review it more in depth in the days to come to get comfortable with the code and let you know if I have some questions. I think you can start building things over the new interface since I have no objection to it.

chukinas commented:
Wonderful! I actually noticed something over the weekend as I was using this in a project. There's a compile warning when a plugin doesn't implement a `field` callback. This wasn't caught in the tests. You can see what I'm talking about if you `use TypedStruct.Plugin` and don't implement a `field/3` or `field/4` callback. I'll find a nice solution to that this week.

It makes me wonder though how we would test for that?

ejpcmac commented on lib/typed_struct.ex:
```diff
@@ -5,6 +5,16 @@ defmodule TypedStruct do
              |> String.split("<!-- MDOC !-->")
              |> Enum.fetch!(1)
 
+  @accumulating_attrs ~w/
+    ts_plugins
+    ts_plugin_fields
+    ts_fields
+    ts_types
+    ts_enforce_keys
+    /a
```
I’m not a big fan of sigils when they are used outside of one-liners. I think it is a little bit harder to read, and their content is not formatted automatically. But please don’t create a commit for this, I’ll do it myself among other tiny cosmetic changes to please my OCD ^^

Otherwise, declaring them as a list here is a pretty neat idea :slightly_smiling_face:

ejpcmac commented on lib/typed_struct.ex:
```diff
@@ -177,31 +170,14 @@ defmodule TypedStruct do
-          )
-        end
-      end)
+    quote bind_quoted: [name: name, type: Macro.escape(type), opts: opts] do
+      TypedStruct.__field__(__ENV__, name, type, opts)
```
Oh, yes, indeed. It seems that I overlooked the `bind_quoted` parameter. But as you did this is crystal clear :) Thank you!

ejpcmac commented:
> I actually noticed something over the weekend as I was using this in a project. There's a compile warning when a plugin doesn't implement a `field` callback.

In fact it even fails to build if you try to define an empty plugin:

```elixir
defmodule ImaginaryPlugin do
  use TypedStruct.Plugin
end
```

chukinas commented:
Quite. Yeah, I'll add a test that checks for that in addition to fixing the bug

ejpcmac commented:
I’m sorry not to have much time at the moment to look deeper into it. Anyway, thank you so much for your work, and the quality of your commits. It’s been really easy to follow them one by one and understand your changes.

chukinas committed:
test: add failing test for empty plugin

This test confirms the problem that an empty plugin doesn't compile,
a problem I hadn't caught before submitting the PR.

Next I'll fix the code to make the test pass.

chukinas committed:
bugfix: empty plugins now compile

Before, empty plugins injected an overrideable field/4 that would call
to field/3 if it existed. This worked fine if there actually was a
field/3. But if there wasn't, you'd get a compile error.

I've fixed that now with a before_compile callback. The code checks for the
existence of field/3 and /4, then inject the correct field/4.

chukinas commented:
I appreciate your comments! Very happy to help.

Ok, I think it's all working now. The test confirms that empty plugins compile. It's still backwards-compatible with `field/3` and if the user defines both `field/3` and `field/4`, they'll get an `IO.warn` saying their `field/3` will be ignored.

ejpcmac commented on lib/typed_struct.ex:
```diff
@@ -177,31 +170,14 @@ defmodule TypedStruct do
 
   @doc false
-  def __field__(mod, name, type, opts) when is_atom(name) do
+  def __field__(%Macro.Env{module: mod} = env, name, type, opts)
```
Shouldn’t we put `env` last for consistency with the plugin callback? I can take this point.

chukinas replied:
Good point. I'm ok with either.

ejpcmac commented on lib/typed_struct/plugin.ex:
```diff
@@ -148,33 +149,76 @@ defmodule TypedStruct.Plugin do
+          Module.defines?(env.module, {:field, 4}, :def)} do
+      {true, true} ->
+        IO.warn([
+          env.module,
```
It would be `Atom.to_string(env.module)`. The current implementation leads to an error when taking this path. I’ll fix this today and add a test for the warning.

chukinas replied:
Thanks!

ejpcmac reviewed (approved):
Again, thank you so much for your changes. There is just one tiny bug when trying to print the warning when both `field/3` and `field/4` are implemented, but I’ll fix this on my side.

I’ll merge it and do a few consistency changes, but the API won’t change. Also, I think it would be worth to hard-deprecate `field/3`, to prepare for v1.0.0 which will not include it. Do you see any objection?

I’ll prepare a v0.3.0 release today, also updating the plugins I own.

chukinas commented:
Hard deprecation of `field/3` sounds good to me :+1: 

Thanks again for being so receptive! This has been fun. Definitely looking forward to that next release.
